coprocessor/fhevm-engine/host-listener/src/poller/mod.rs

```diff
@@ -0,0 +1,350 @@
+                error = %err.error,
+                retries = err.retries,
+                "Failed to fetch chain id after retries"
+            );
+            sleep(config.retry_interval).await;
+            return Ok(());
+        }
+    };
+    let chain_id_str = chain_id.to_string();
+    if http_retries > 0 {
+        inc_http_retries(&chain_id_str, http_retries);
+    }
+    blockchain_timeout_tick.update();
+
+    let mut db = Database::new(
+        &config.database_url,
+        &config.coprocessor_api_key,
+        DEFAULT_DEPENDENCE_CACHE_SIZE,
+    )
+    .await?;
+
+    if chain_id != db.chain_id {
+        error!(
+            chain_id_blockchain = ?chain_id,
+            chain_id_db = ?db.chain_id,
+            tenant_id = ?db.tenant_id,
+            coprocessor_api_key = ?config.coprocessor_api_key,
+            "Chain ID mismatch with database",
+        );
+        return Err(anyhow!(
+            "Chain ID mismatch with database, blockchain: {} vs db: {}, tenant_id: {}, coprocessor_api_key: {}",
+            chain_id,
+            db.chain_id,
+            db.tenant_id,
+            config.coprocessor_api_key
+        ));
+    }
+
+    let initial_anchor =
+        db.poller_get_last_caught_up_block(chain_id as i64).await?;
+    db.tick.update();
+    let mut last_caught_up_block = match initial_anchor {
+        Some(block) => u64::try_from(block)
+            .context("last_caught_up_block cannot be negative")?,
+        None => {
+            let initial = db.read_last_valid_block().await.unwrap_or(0);
+            db.poller_set_last_caught_up_block(chain_id as i64, initial)
+                .await?;
+            db.tick.update();
+            u64::try_from(initial)
+                .context("initial last_caught_up_block cannot be negative")?
+        }
+    };
+
+    let health_check = HealthCheck {
+        blockchain_timeout_tick: blockchain_timeout_tick.clone(),
+        blockchain_tick: blockchain_tick.clone(),
+        blockchain_provider: blockchain_provider.clone(),
+        database_pool: db.pool.clone(),
+        database_tick: db.tick.clone(),
+    };
+    let health_check_cancel_token = CancellationToken::new();
+    let health_check_server = HealthHttpServer::new(
+        Arc::new(health_check),
+        config.health_port,
+        health_check_cancel_token.clone(),
+    );
+    tokio::spawn(async move {
+        if let Err(err) = health_check_server.start().await {
+            error!(error = %err, "Health check server failed");
+        }
+    });
+
+    info!(
+        chain_id = chain_id,
+        last_caught_up_block = last_caught_up_block,
+        finality_lag = config.finality_lag,
+        batch_size = config.batch_size,
+        poll_interval_ms = config.poll_interval.as_millis(),
+        retry_interval_ms = config.retry_interval.as_millis(),
+        "Starting host-listener poller"
+    );
+
+    loop {
+        let (latest, latest_retries) = match client.latest_block_number().await
```

Comment:
what happens if the client gets invalid ? Can the provider be in a bad state at some point ?